processor/src/range/aux_trace.rs

```diff
@@ -0,0 +1,235 @@
+        let mut acc = E::ONE;
+        for (row_idx, &v) in v_col
+            .iter()
+            .enumerate()
+            .take(v_col.len() - 1)
+            .skip(self.start_16bit)
+        {
+            // This is the 16-bit section, where the running product must be reduced.
+            let v_next = v_col[row_idx + 1].into();
+            let value = alpha + v_next - v.into();
+
+            // Accumulate the transition difference values by which the running product must be reduced.
+            diff_values.push(value);
+
+            // Accumulate the product of the differences.
+            if value != E::ZERO {
+                acc *= value;
+            }
+        }
+
+        // Invert the accumulated product and multiply it by the result from the 8-bit section.
+        acc = acc.inv() * aux_column[self.start_16bit];
+
+        // Do a modified version of batch inversion. We don't actually want an array of inverted
+        // diff_values [1/a, 1/b, 1/c, ...], we want an array of inverted products all of which are
+        // multiplied by the same 8-bit result `res`, e.g. [res/a, res/ab, res/abc, ...].
+        for idx in (0..diff_values.len()).rev() {
+            aux_column[self.start_16bit + idx + 1] = acc;
+            if diff_values[idx] != E::ZERO {
+                acc *= diff_values[idx];
+            }
+        }
+
+        aux_column
+    }
+
+    /// Builds the execution trace of the range check `p1` and `q` columns which ensure that the
+    /// range check lookups performed by user operations match those executed by the Range Checker.
+    fn build_aux_col_p1<E: FieldElement<BaseField = Felt>>(
+        &self,
+        main_trace: &Matrix<Felt>,
+        alphas: &[E],
+    ) -> (Vec<E>, Vec<E>) {
+        // compute the inverses for range checks performed by operations.
+        let (_, inv_row_values) =
+            build_lookup_table_row_values(&self.cycle_range_check_values(), alphas);
+
+        // allocate memory for the running product column and set the initial value to ONE
+        let mut q = unsafe { uninit_vector(main_trace.num_rows()) };
+        let mut p1 = unsafe { uninit_vector(main_trace.num_rows()) };
+        q[0] = E::ONE;
+        p1[0] = E::ONE;
+
+        // keep track of the last updated row in the `p1` running product column. the `q` column
+        // index is always one row behind, since `q` is filled with intermediate values in the same
+        // row as the operation is executed, whereas `p1` is filled with result values that are
+        // added to the next row after the operation's execution.
+        let mut p1_idx = 0;
+        // keep track of the next row to be included from the user op range check values.
+        let mut rc_user_op_idx = 0;
+
+        // the first half of the trace only includes values from the operations.
+        for (clk, range_checks) in self
+            .cycle_range_checks
+            .iter()
+            .take_while(|(&step, _)| step <= self.start_16bit)
```

Comment:
Is it not possible to use [BTreeMap::range()](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range) here?

Reply:
It is, thank you!